app/ts/Elicitation/ElicitationUtil.ts

```diff
@@ -0,0 +1,136 @@
+import IInputCriterion from './Interface/IInputCriterion';
+import IOrdinalRanking from './Interface/IOrdinalRanking';
+import IRankingAnswer from './Interface/IRankingAnswer';
+
+export function buildElicitationCriteria(
+  input: IInputCriterion[]
+): Record<string, IElicitationCriterion> {
+  return _(input)
+    .map((criterion: IInputCriterion) => {
+      const elicitationCriterion: IElicitationCriterion = {
+        mcdaId: criterion.id,
+        title: criterion.title,
+        scales: [criterion.worst, criterion.best],
+        unitOfMeasurement: criterion.dataSources[0].unitOfMeasurement.label,
+        pvfDirection: criterion.dataSources[0].pvf.direction,
+        description: criterion.description
+      };
+      return [criterion.id, elicitationCriterion];
+    })
+    .fromPairs()
+    .value();
+}
+
+export function getWorst(criterion: IElicitationCriterion): number {
+  if (criterion.scales) {
+    return significantDigits(
+      criterion.pvfDirection === 'increasing'
+        ? Math.min(...criterion.scales)
+        : Math.max(...criterion.scales)
+    );
+  } else {
+    return -1;
+  }
+}
+
+export function getBest(criterion: IElicitationCriterion): number {
+  if (criterion.scales) {
+    return significantDigits(
+      criterion.pvfDirection === 'increasing'
+        ? Math.max(...criterion.scales)
+        : Math.min(...criterion.scales)
+    );
+  } else {
+    return -1;
+  }
+}
+
+export function findCriterionIdForRank(
+  criteria: Record<string, IElicitationCriterion>,
+  rankings: Record<string, IRankingAnswer>,
+  rank: number
+): string {
+  return _.find(criteria, (criterion) => {
+    return rankings[criterion.mcdaId].rank === rank;
+  }).mcdaId;
+}
+
+export function assignMissingRankings(
+  rankings: Record<string, IRankingAnswer>,
+  selectedCriterionId: string,
+  rank: number,
+  criteria: Record<string, IElicitationCriterion>
+): Record<string, IRankingAnswer> {
+  let finishedRankings = _.cloneDeep(rankings);
+  const secondToLastRanking = buildRankingAnswer(selectedCriterionId, rank);
+  finishedRankings[selectedCriterionId] = secondToLastRanking;
+  const lastCriterionId = findCriterionIdWithoutRanking(
+    criteria,
+    finishedRankings
+  );
+  const lastRanking = buildRankingAnswer(lastCriterionId, rank + 1);
+  finishedRankings[lastCriterionId] = lastRanking;
+  return finishedRankings;
+}
+
+function findCriterionIdWithoutRanking(
+  criteria: Record<string, IElicitationCriterion>,
+  rankings: Record<string, IRankingAnswer>
+): string {
+  return _.find(criteria, (criterion) => {
+    return (
+      rankings[criterion.mcdaId] === undefined ||
+      rankings[criterion.mcdaId].rank === UNRANKED
+    );
+  }).mcdaId;
+}
+
+function buildRankingAnswer(criterionId: string, rank: number): IRankingAnswer {
+  return {
+    criterionId: criterionId,
+    rank: rank
+  };
+}
+
+export function buildOrdinalPreferences(
+  answers: IRankingAnswer[]
```

Comment:
Wouldn't it be easier to use a Record<string, IRankingAnswer> keyed by criterionId? Then you don't have to sort and you can just do a _.map

Reply:
It seems that using an array is easier as you have to use idx+1 ...

Reply:
I feel like using indices makes the code less readible, and is more error prone.